azure/services/privateendpoints/privateendpoints_test.go

```diff
@@ -21,114 +21,185 @@ import (
-		CustomNetworkInterfaceName:    "pestaticconfig",
-		PrivateIPAddresses:            []string{"10.0.0.1"},
+	fakeASOPrivateEndpointWithManualPrivateLinkService = asonetworkv1.PrivateEndpoint{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fakePrivateEndpoint.Name,
+			Namespace: fakePrivateEndpoint.Namespace,
+		},
+		Spec: asonetworkv1.PrivateEndpoint_Spec{
+			ApplicationSecurityGroups: []asonetworkv1.ApplicationSecurityGroupSpec_PrivateEndpoint_SubResourceEmbedded{
+				{
+					Reference: &genruntime.ResourceReference{
+						ARMID: fakePrivateEndpoint.ApplicationSecurityGroups[0],
+					},
+				},
+				{
+					Reference: &genruntime.ResourceReference{
+						ARMID: fakePrivateEndpoint.ApplicationSecurityGroups[1],
+					},
+				},
+			},
+			AzureName: fakePrivateEndpoint.Name,
+			ManualPrivateLinkServiceConnections: []asonetworkv1.PrivateLinkServiceConnection{{
+				Name: ptr.To(fakePrivateEndpoint.PrivateLinkServiceConnections[0].Name),
+				PrivateLinkServiceReference: &genruntime.ResourceReference{
+					ARMID: fakePrivateEndpoint.PrivateLinkServiceConnections[0].PrivateLinkServiceID,
+				},
+				GroupIds:       fakePrivateEndpoint.PrivateLinkServiceConnections[0].GroupIDs,
+				RequestMessage: ptr.To(fakePrivateEndpoint.PrivateLinkServiceConnections[0].RequestMessage),
+			}},
+			Owner: &genruntime.KnownResourceReference{
+				Name: fakePrivateEndpoint.ResourceGroup,
+			},
+			Subnet: &asonetworkv1.Subnet_PrivateEndpoint_SubResourceEmbedded{
+				Reference: &genruntime.ResourceReference{
+					ARMID: fakePrivateEndpoint.SubnetID,
+				},
+			},
+			IpConfigurations: []asonetworkv1.PrivateEndpointIPConfiguration{
+				{
+					PrivateIPAddress: ptr.To(fakePrivateEndpoint.PrivateIPAddresses[0]),
+				},
+				{
+					PrivateIPAddress: ptr.To(fakePrivateEndpoint.PrivateIPAddresses[1]),
+				},
+			},
+			CustomNetworkInterfaceName: ptr.To(fakePrivateEndpoint.CustomNetworkInterfaceName),
+			Location:                   ptr.To(fakePrivateEndpoint.Location),
+			Tags:                       map[string]string{"sigs.k8s.io_cluster-api-provider-azure_cluster_test_cluster": "owned", "Name": "test_private_endpoint_1", "test_tag1": "test_value1", "test_tag2": "test_value2"},
+		},
 	}
 
 	emptyPrivateEndpointSpec = PrivateEndpointSpec{}
-	fakePrivateEndpointSpecs = []azure.ResourceSpecGetter{&fakePrivateEndpoint1, &fakePrivateEndpoint2, &fakePrivateEndpoint3, &emptyPrivateEndpointSpec}
+	fakePrivateEndpointSpecs = []azure.ASOResourceSpecGetter[*asonetworkv1.PrivateEndpoint]{&fakePrivateEndpoint, &fakePrivateEndpointWithManualApproval, &emptyPrivateEndpointSpec}
 
 	internalError = autorest.NewErrorWithResponse("", "", &http.Response{StatusCode: http.StatusInternalServerError}, "Internal Server Error")
 	notDoneError  = azure.NewOperationNotDoneError(&infrav1.Future{})
 )
 
+func init() {
+	_ = clusterv1.AddToScheme(scheme.Scheme)
+	_ = asonetworkv1.AddToScheme(scheme.Scheme)
+}
+
 func TestReconcilePrivateEndpoint(t *testing.T) {
```

Comment:
I still think it's not worth testing the generic `Reconcile` implementation in every service that uses it. If there are gaps in the tests for the generic service, then we should fill those there.

Reply:
I agree that testing the `Reconcile` functionality is redundant. 
But these are good tests to test the overall code flow; i.e. `specs` -> `reconcile` -> `createOrUpdate` -> `updateStatus` a service. 
It was through these tests I discovered that `PrivateEndpointSpecs` would never return `nil` `[]azure.ASOResourceSpecGetter[*asonetworkv1.PrivateEndpoint]`, instead it would return an empty list. 

(In our scenario, a `nil` list and an `empty` list doesn't really make a difference since we only `range` over them in _aso/service.go_

Reply:
privateendpoints.go isn't providing any logic itself outside of initializing a struct, and those struct fields are all generic enough that any tests here could be replicated in the generic tests. Especially if we keep replicating these tests across all the services, I think the maintenance cost will pretty quickly surpass how many new bugs we find. Before it was more necessary since each service had its own implementation for these functions, but these tests are executing literally the same code that's already tested.

Reply:
Dropped the tests via 178d32f4039b9117767f49c599edcaf0bb1c202a